proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1565 @@
+    - `{key}` — the key portion of the map item (or index if an array item or document).
+    - `{value}` — the value of the annotated node (supplied as [`repr(value)`](https://github.com/google/starlark-go/blob/bb14e15/doc/spec.md#repr))
+    - `{desc}` — the description supplied in the 0th item in this rule.
+    - `{failure}` — when `assertion` fails, the message from that `assert.fail()` or empty string.
+- `<named-rules>` — any of the keyword arguments described in [Named Rules](#named-rules), below.
+- `when=` (`function(value) : None` | `function(value) : bool`) — criteria for when the validation rules should be checked. If the criteria is met (function either returns `True` or `None`), the validations are checked; otherwise (function either returns `False` or `assert.fail()`s, none of the validations are checked.
+- `when_null_skip=` (`bool`) — a special-case of `when=` that checks if the value of the annotated node is `null`.
+
+Each rule is evaluated, in the order they appear in the annotation (left-to-right):
+- if all rules pass (either returns `True` or `None`), then the value is valid.
+- if a rule returns `False` (not `None`) or `assert.fail()`s, then the valid is invalid.
+
+The default `message` is: 
+```
+\"{key}\" requires a valid value: {desc}; {failure}.
+```
+- centering around the word "...requires..." helps the message work both in the case where the consumer supplied no value (see [Use Case: Required input](#use-case-required-input)) _and_ when they supplied an invalid value.
+- `{value}` is omitted to not contribute to the risk of leaking secrets.
+
+> **Note** \
+> Examples below illustrate how an author would specify a validation (using `@schema/validation`) that becomes an `@assert/validate` as this is the driving use case at this time.
+
+_Example 1: Named rules_
+
+```yaml
+#@data/values-schema
+---
+#@schema/validation min=49142 max=65535
+adminPort: 1024
+```
+which yields a schema base document:
+```yaml
+---
+#@assert/validate min=49142 max=65535
+adminPort: 1024
+```
+
+_Example 2: Custom assertion-based rule_
+
+```yaml
+#@data/values-schema
+---
+#@ def is_dynamic_port(n):
+#@   n >= 49142 and n <= 65535 or assert.fail("{} is not in the dynamic port range".format(n))
+#@ end
+
+#@schema/validation ("a TCP/IP port in the \"dynamic\" range: 49142 and 65535, inclusive", is_valid_port)
+adminPort: 1024
+```
+which yields a schema base document:
+```yaml
+---
+#@assert/validate ("a TCP/IP port in the \"dynamic\" range: 49142 and 65535, inclusive", is_valid_port)
+adminPort: 1024
+```
+
+Yields:
+> "adminPort" requires a valid value: a TCP/IP port in the "dynamic" range: 49142 and 65535, inclusive; 1024 is not in the dynamic port range.
+
+_Example 3: Custom violation message_
+
+```yaml
+#@data/values-schema
+---
+#@ def is_dynamic_port(n):
```

Comment:
Is this function name a typo? The example uses is_valid_port as the assertion name

Reply:
👍🏻